### ISSUE TITLE
check_http: segfault with monitoring-plugins-2.3.3

### ISSUE THREAD

**Issue by hydrapolic** [monitoring-plugins/monitoring-plugins#1836]
After upgrading from monitoring-plugins 2.3.2 -> 2.3.3 I'm getting segfaults with check_http:

```
# /usr/lib64/nagios/plugins/check_http -I localhost -H site.tld
Segmentation fault

# /usr/lib64/nagios/plugins/check_http -I localhost -H site.tld -C 10
OK - Certificate '*.tld' will expire on Thu 28 Sep 2023 11:59:59 PM GMT +0000.

# /usr/lib64/nagios/plugins/check_curl -I localhost -H site.tld
HTTP OK: HTTP/1.1 301 Moved Permanently - 290 bytes in 0.002 second response time |time=0.001838s;;;0.000000;10.000000 size=290B;;;0;
```

dmesg:
```
Feb  5 01:06:26 srv kernel: [ 2637.301495] check_http[1950]: segfault at 0 ip 000055cda592b380 sp 00007ffc7ea8ff70 error 6 in check_http[55cda592a000+e000]
Feb  5 01:06:26 srv kernel: [ 2637.301555] Code: 40 00 48 89 44 24 10 48 89 c1 48 ff c0 80 78 ff 0d 75 ef 4c 8d 71 02 4e 8d 64 39 04 4d 85 ff 75 a4 66 0f 1f 84 00 00 00 00 00 <c6> 04 2b 00 48 8b 44 24 18 64 48 2b 04 25 28 00 00 00 75 6d 48 83
```

strace:
```
execve("/usr/lib64/nagios/plugins/check_http", ["/usr/lib64/nagios/plugins/check_"..., "-I", "localhost", "-H", "site.tld"], 0x7ffc856e3df0 /* 26 vars */) = 0
brk(NULL)                               = 0x55f6ce662000
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6ee278a000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=22916, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 22916, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f6ee2784000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib64/libssl.so.1.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0\0\0\0\0\0\0\0"..., 832) = 832
newfstatat(3, "", {st_mode=S_IFREG|0755, st_size=593160, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 595792, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f6ee26f2000
mprotect(0x7f6ee270f000, 425984, PROT_NONE) = 0
mmap(0x7f6ee270f000, 315392, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1d000) = 0x7f6ee270f000
mmap(0x7f6ee275c000, 106496, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x6a000) = 0x7f6ee275c000
mmap(0x7f6ee2777000, 53248, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x84000) = 0x7f6ee2777000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib64/libcrypto.so.1.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0\0\0\0\0\0\0\0"..., 832) = 832
newfstatat(3, "", {st_mode=S_IFREG|0755, st_size=2848648, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 2868240, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f6ee2435000
mmap(0x7f6ee24a9000, 1613824, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x74000) = 0x7f6ee24a9000
mmap(0x7f6ee2633000, 577536, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1fe000) = 0x7f6ee2633000
mmap(0x7f6ee26c0000, 184320, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x28a000) = 0x7f6ee26c0000
mmap(0x7f6ee26ed000, 17424, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f6ee26ed000
close(3)                                = 0
openat(AT_FDCWD, "/lib64/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\2003\2\0\0\0\0\0"..., 832) = 832
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
newfstatat(3, "", {st_mode=S_IFREG|0755, st_size=1856312, ...}, AT_EMPTY_PATH) = 0
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
mmap(NULL, 1887536, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f6ee2268000
mmap(0x7f6ee228a000, 1351680, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x22000) = 0x7f6ee228a000
mmap(0x7f6ee23d4000, 339968, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x16c000) = 0x7f6ee23d4000
mmap(0x7f6ee2427000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1be000) = 0x7f6ee2427000
mmap(0x7f6ee242d000, 32048, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f6ee242d000
close(3)                                = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6ee2266000
arch_prctl(ARCH_SET_FS, 0x7f6ee2266b80) = 0
set_tid_address(0x7f6ee2266e50)         = 1063
set_robust_list(0x7f6ee2266e60, 24)     = 0
rseq(0x7f6ee22674a0, 0x20, 0, 0x53053053) = 0
mprotect(0x7f6ee2427000, 16384, PROT_READ) = 0
mprotect(0x7f6ee26c0000, 176128, PROT_READ) = 0
mprotect(0x7f6ee2777000, 32768, PROT_READ) = 0
mprotect(0x55f6ccde2000, 4096, PROT_READ) = 0
mprotect(0x7f6ee27bd000, 8192, PROT_READ) = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
munmap(0x7f6ee2784000, 22916)           = 0
getrandom("\x27\xd2\xf8\xee\x33\x67\x5c\xbe", 8, GRND_NONBLOCK) = 8
brk(NULL)                               = 0x55f6ce662000
brk(0x55f6ce683000)                     = 0x55f6ce683000
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=3382112, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 3382112, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f6ee1f2c000
close(3)                                = 0
rt_sigaction(SIGALRM, {sa_handler=0x55f6ccdd5340, sa_mask=[ALRM], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f6ee229f7a0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
alarm(10)                               = 0
newfstatat(AT_FDCWD, "/etc/nsswitch.conf", {st_mode=S_IFREG|0644, st_size=2091, ...}, 0) = 0
newfstatat(AT_FDCWD, "/", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
openat(AT_FDCWD, "/etc/nsswitch.conf", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=2091, ...}, AT_EMPTY_PATH) = 0
read(3, "#\n# /etc/nsswitch.conf\n#\n# An ex"..., 4096) = 2091
read(3, "", 4096)                       = 0
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=2091, ...}, AT_EMPTY_PATH) = 0
close(3)                                = 0
newfstatat(AT_FDCWD, "/etc/resolv.conf", {st_mode=S_IFREG|0644, st_size=140, ...}, 0) = 0
openat(AT_FDCWD, "/etc/host.conf", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=825, ...}, AT_EMPTY_PATH) = 0
read(3, "# /etc/host.conf:\n\n# The  file /"..., 4096) = 825
read(3, "", 4096)                       = 0
close(3)                                = 0
futex(0x7f6ee243442c, FUTEX_WAKE_PRIVATE, 2147483647) = 0
openat(AT_FDCWD, "/etc/resolv.conf", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=140, ...}, AT_EMPTY_PATH) = 0
read(3, "# Generated by netifrc for inter"..., 4096) = 140
read(3, "", 4096)                       = 0
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=140, ...}, AT_EMPTY_PATH) = 0
close(3)                                = 0
openat(AT_FDCWD, "/etc/hosts", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=1265, ...}, AT_EMPTY_PATH) = 0
lseek(3, 0, SEEK_SET)                   = 0
read(3, "# /etc/hosts: Local Host Databas"..., 4096) = 1265
read(3, "", 4096)                       = 0
close(3)                                = 0
openat(AT_FDCWD, "/etc/gai.conf", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=2584, ...}, AT_EMPTY_PATH) = 0
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=2584, ...}, AT_EMPTY_PATH) = 0
read(3, "# Configuration for getaddrinfo("..., 4096) = 2584
read(3, "", 4096)                       = 0
close(3)                                = 0
futex(0x7f6ee24332e4, FUTEX_WAKE_PRIVATE, 2147483647) = 0
socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 3
bind(3, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = 0
getsockname(3, {sa_family=AF_NETLINK, nl_pid=1063, nl_groups=00000000}, [12]) = 0
sendto(3, [{nlmsg_len=20, nlmsg_type=0x16 /* NLMSG_??? */, nlmsg_flags=NLM_F_REQUEST|0x300, nlmsg_seq=1675555504, nlmsg_pid=0}, "\x00\x00\x00\x00"], 20, 0, {sa_family=AF_NETLINK, nl_pid=0, nl_groups
=00000000}, 12) = 20
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[[{nlmsg_len=84, nlmsg_type=0x14 /* NLMSG_??? */, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=167
5555504, nlmsg_pid=1063}, "\x02\x08\x80\xfe\x01\x00\x00\x00\x08\x00\x01\x00\x7f\x00\x00\x01\x08\x00\x02\x00\x7f\x00\x00\x01\x08\x00\x04\x00\x7f\xff\xff\xff"...], [{nlmsg_len=88, nlmsg_type=0x14 /* N
LMSG_??? */, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1675555504, nlmsg_pid=1063}, "\x02\x18\x80\x00\x02\x00\x00\x00\x08\x00\x01\x00\x0a\x14\x07\x8d\x08\x00\x02\x00\x0a\x14\x07\x8d\x08\x00\x04\x00\x0a\x14
\x07\xff"...], [{nlmsg_len=88, nlmsg_type=0x14 /* NLMSG_??? */, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1675555504, nlmsg_pid=1063}, "\x02\x10\x80\x00\x04\x00\x00\x00\x08\x00\x01\x00\xac\x11\x00\x01\x08\
x00\x02\x00\xac\x11\x00\x01\x08\x00\x04\x00\xac\x11\xff\xff"...]], iov_len=4096}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 260
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[[{nlmsg_len=72, nlmsg_type=0x14 /* NLMSG_??? */, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=167
5555504, nlmsg_pid=1063}, "\x0a\x80\x80\xfe\x01\x00\x00\x00\x14\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x14\x00\x06\x00"...], [{nlmsg_len=72, nlmsg_type=0x14 /* N
LMSG_??? */, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1675555504, nlmsg_pid=1063}, "\x0a\x40\x80\xfd\x02\x00\x00\x00\x14\x00\x01\x00\xfe\x80\x00\x00\x00\x00\x00\x00\x02\x50\x56\xff\xfe\x80\x82\x15\x14\x00
\x06\x00"...], [{nlmsg_len=72, nlmsg_type=0x14 /* NLMSG_??? */, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1675555504, nlmsg_pid=1063}, "\x0a\x40\x80\xfd\x04\x00\x00\x00\x14\x00\x01\x00\xfe\x80\x00\x00\x00\
x00\x00\x00\x00\x42\x30\xff\xfe\xd5\x05\x3e\x14\x00\x06\x00"...], [{nlmsg_len=72, nlmsg_type=0x14 /* NLMSG_??? */, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1675555504, nlmsg_pid=1063}, "\x0a\x40\x80\xfd\x
06\x00\x00\x00\x14\x00\x01\x00\xfe\x80\x00\x00\x00\x00\x00\x00\x58\x47\x02\xff\xfe\x27\xef\x81\x14\x00\x06\x00"...]], iov_len=4096}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 288
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[{nlmsg_len=20, nlmsg_type=NLMSG_DONE, nlmsg_flags=NLM_F_MULTI, nlmsg_seq=1675555504, nl
msg_pid=1063}, 0], iov_len=4096}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 20
close(3)                                = 0
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 3
connect(3, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
getsockname(3, {sa_family=AF_INET, sin_port=htons(14495), sin_addr=inet_addr("127.0.0.1")}, [28 => 16]) = 0
close(3)                                = 0
socket(AF_INET6, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 3
connect(3, {sa_family=AF_INET6, sin6_port=htons(80), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::1", &sin6_addr), sin6_scope_id=0}, 28) = 0
getsockname(3, {sa_family=AF_INET6, sin6_port=htons(31411), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::1", &sin6_addr), sin6_scope_id=0}, [28]) = 0
close(3)                                = 0
socket(AF_INET6, SOCK_STREAM, IPPROTO_TCP) = 3
connect(3, {sa_family=AF_INET6, sin6_port=htons(80), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::1", &sin6_addr), sin6_scope_id=0}, 28) = -1 ECONNREFUSED (Connection refused)
close(3)                                = 0
socket(AF_INET, SOCK_STREAM, IPPROTO_TCP) = 3
connect(3, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
sendto(3, "GET / HTTP/1.1\r\nUser-Agent: chec"..., 118, 0, NULL, 0) = 118
read(3, "HTTP/1.1 301 Moved Permanently\r\n"..., 8191) = 291
read(3, "", 8191)                       = 0
close(3)                                = 0
openat(AT_FDCWD, "/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=2998, ...}, AT_EMPTY_PATH) = 0
read(3, "# Locale name alias data base.\n#"..., 4096) = 2998
read(3, "", 4096)                       = 0
close(3)                                = 0
openat(AT_FDCWD, "/usr/share/locale/en_US.utf8/LC_MESSAGES/monitoring-plugins.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en_US/LC_MESSAGES/monitoring-plugins.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en.utf8/LC_MESSAGES/monitoring-plugins.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en/LC_MESSAGES/monitoring-plugins.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
alarm(0)                                = 10
openat(AT_FDCWD, "/usr/lib64/gconv/gconv-modules.cache", O_RDONLY) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=27012, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 27012, PROT_READ, MAP_SHARED, 3, 0) = 0x7f6ee1f25000
close(3)                                = 0
futex(0x7f6ee242ca4c, FUTEX_WAKE_PRIVATE, 2147483647) = 0
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=NULL} ---
+++ killed by SIGSEGV +++
```

gdb:
```
# gdb /usr/lib64/nagios/plugins/check_http
GNU gdb (Gentoo 12.1 vanilla) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/lib64/nagios/plugins/check_http...
(gdb) run -I localhost -H site.tld
Starting program: /usr/lib64/nagios/plugins/check_http -I localhost -H site.tld
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
unchunk_content (content=content@entry=0x55555557e55e "0\r\n\r\n") at check_http.c:1465
1465    check_http.c: No such file or directory.
(gdb) bt
#0  unchunk_content (content=content@entry=0x55555557e55e "0\r\n\r\n") at check_http.c:1465
#1  0x0000555555558c94 in check_http () at check_http.c:1286
#2  0x0000555555557982 in main (argc=<optimized out>, argv=<optimized out>) at check_http.c:186
```

Gentoo-bug: https://bugs.gentoo.org/893252

**Comment by RincewindsHat** [comment]
hm, my fault then. gonna take a look at that.

**Comment by hydrapolic** [comment]
> hm, my fault then. gonna take a look at that.

Thank you @RincewindsHat

**Comment by RincewindsHat** [comment]
Should be fixed with 6d3e44d2d8395076060e9c741e9b173dc5d57b76

**Comment by hydrapolic** [comment]
It works on my server, thanks @RincewindsHat!

**Comment by RincewindsHat** [comment]
This is should be fixed with #1840
